zebra-consensus/src/block/subsidy.rs

```diff
@@ -0,0 +1,228 @@
+//! Calculation of Block Subsidy, Funding Streams, and Founders’ Reward - [§7.7][7.7]
+//!
+//! [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+
+use std::convert::TryFrom;
+
+use zebra_chain::{
+    amount::{Amount, Error, NonNegative},
+    block::Height,
+    parameters::{Network, NetworkUpgrade::*},
+};
+
+use crate::parameters::subsidy::*;
+
+/// `SlowStartShift()` as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+fn slow_start_shift() -> Height {
+    Height(SLOW_START_INTERVAL.0 / 2)
+}
+
+/// `Halving(height)` as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+fn halving_factor(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
+    if height < SLOW_START_INTERVAL {
+        panic!("can't verify before block {}", SLOW_START_INTERVAL.0)
+    }
+    let blossom_height = Blossom
+        .activation_height(network)
+        .expect("blossom activation height should be available");
+    if height >= blossom_height {
+        let scaled_halvings = (blossom_height - slow_start_shift()) as u32
+            * BLOSSOM_POW_TARGET_SPACING_RATIO as u32
+            + height.0
+            - blossom_height.0;
+        Amount::try_from((1 << scaled_halvings / POST_BLOSSOM_HALVING_INTERVAL.0) as i64)
+    } else {
+        Amount::try_from(
+            ((height - slow_start_shift()) as u32 / PRE_BLOSSOM_HALVING_INTERVAL.0) as i64,
+        )
+    }
+}
```

Comment:
The halving isn't a currency amount, it's a divisor.

Also fix up some of the internal types, and refactor that code into multiple variables.

```suggestion
/// The divisor used for halvings.
///
/// `1 << Halving(height)`, as described in [protocol specification §7.7][7.7]
///
/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
fn halving_divisor(height: Height, network: Network) -> u64 {
    if height < SLOW_START_INTERVAL {
        panic!("can't verify before block {}", SLOW_START_INTERVAL.0)
    }
    let blossom_height = Blossom
        .activation_height(network)
        .expect("blossom activation height should be available");
    if height >= blossom_height {
        let scaled_pre_blossom_height = (blossom_height - slow_start_shift()) as u64
            * BLOSSOM_POW_TARGET_SPACING_RATIO;
        let post_blossom_height = (height - blossom_height) as u64;
        let halving_shift = (scaled_pre_blossom_height + post_blossom_height) / (POST_BLOSSOM_HALVING_INTERVAL.0 as u64);
        1 << halving_shift
    } else {
        let scaled_pre_blossom_height = (height - slow_start_shift()) as u64;
        let halving_shift = scaled_pre_blossom_height / (PRE_BLOSSOM_HALVING_INTERVAL.0 as u64);
        1 << halving_shift
    }
}
```

Reply:
Applied at https://github.com/ZcashFoundation/zebra/pull/1051/commits/364f70eea3f19be9c35d884d861b5096018aa4fd